presto-main/src/main/java/com/facebook/presto/metadata/PolymorphicScalarFunctionBuilder.java

```diff
@@ -220,56 +195,140 @@ public MethodsGroupBuilder methods(List<String> methodNames)
             requireNonNull(methodNames, "methodNames is null");
             checkArgument(!methodNames.isEmpty(), "methods list is empty");
 
-            List<Method> matchingMethods = asList(clazz.getMethods()).stream()
+            List<MethodAndNativeContainerTypes> matchingMethods = asList(clazz.getMethods()).stream()
                     .filter(method -> methodNames.contains(method.getName()))
+                    .map(method -> new MethodAndNativeContainerTypes(method, nCopies(signature.getArgumentTypes().size(), Optional.empty())))
                     .collect(toImmutableList());
             List<String> matchingMethodNames = matchingMethods.stream()
-                    .map(Method::getName)
+                    .map(methodAndNativeContainerTypes -> methodAndNativeContainerTypes.getMethod().getName())
                     .collect(toImmutableList());
 
             for (String methodName : methodNames) {
                 checkState(matchingMethodNames.contains(methodName), "method %s was not found in %s", methodName, clazz);
             }
 
-            this.methods = matchingMethods;
+            this.methodAndNativeContainerTypes = matchingMethods;
             return this;
         }
 
         public MethodsGroupBuilder withExtraParameters(Function<SpecializeContext, List<Object>> extraParametersFunction)
         {
-            checkState(methods != null, "methods must be selected first");
+            checkState(methodAndNativeContainerTypes != null, "methods must be selected first");
             requireNonNull(extraParametersFunction, "extraParametersFunction is null");
             this.extraParametersFunction = Optional.of(extraParametersFunction);
             return this;
         }
 
+        public MethodsGroupBuilder methodWithExplicitJavaTypes(String methodName, List<Optional<Class<?>>> types)
+        {
+            Optional<MethodAndNativeContainerTypes> methodAndNativeContainerTypesOptional = this.methodAndNativeContainerTypes.stream()
+                    .filter(methodAndNativeContainerTypes -> methodAndNativeContainerTypes.getMethod().getName().equals(methodName))
+                    .findFirst();
+            checkState(methodAndNativeContainerTypesOptional.isPresent(), "method needs to be defined first");
+            int argumentSize = signature.getArgumentTypes().size();
+            checkState(types.size() == argumentSize, "not matching number of arguments: %s (should have %s)", types.size(), argumentSize);
+            methodAndNativeContainerTypesOptional.get().setExplicitNativeContainerTypes(types);
+            return this;
```

Comment:
Can you validate here that `types` is specified if and only if the corresponding position is `BLOCK_AND_POSITION`?